src/generator/models.ts

```diff
@@ -423,3 +287,232 @@ function generateParamGroupStruct(lang: Language, params: Parameter[]): StructDe
+    } else {
+      marshaller += `\t\tobjectMap["${prop.serializedName}"] = ${receiver}.${prop.language.go!.name}\n`;
+    }
+    marshaller += `\t}\n`;
+  }
+  marshaller += '\treturn json.Marshal(objectMap)\n';
+  marshaller += '}\n\n';
+  structDef.Methods.push({ name: 'MarshalJSON', text: marshaller });
+}
+
+function generateDiscriminatedTypeUnmarshaller(obj: ObjectSchema, structDef: StructDef, parentType?: ObjectSchema) {
+  const typeName = structDef.Language.name;
+  const receiver = typeName[0].toLowerCase();
+  let unmarshaller = `func (${receiver} *${typeName}) UnmarshalJSON(data []byte) error {\n`;
+  // polymorphic type, or type containing a polymorphic type
+  unmarshaller += '\tvar rawMsg map[string]*json.RawMessage\n';
+  unmarshaller += '\tif err := json.Unmarshal(data, &rawMsg); err != nil {\n';
+  unmarshaller += '\t\treturn err\n';
+  unmarshaller += '\t}\n';
+  unmarshaller += '\tfor k, v := range rawMsg {\n';
+  unmarshaller += '\t\tvar err error\n';
+  unmarshaller += '\t\tswitch k {\n';
+  // unmarshal each field one by one
+  for (const prop of values(structDef.Properties)) {
+    unmarshaller += `\t\tcase "${prop.serializedName}":\n`;
+    unmarshaller += '\t\t\tif v != nil {\n';
+    if (prop.schema.language.go!.discriminatorInterface) {
+      unmarshaller += `\t\t\t\t${receiver}.${prop.language.go!.name}, err = unmarshal${prop.schema.language.go!.discriminatorInterface}(*v)\n`;
+    } else if (isArraySchema(prop.schema) && prop.schema.elementType.language.go!.discriminatorInterface) {
+      unmarshaller += `\t\t\t\t${receiver}.${prop.language.go!.name}, err = unmarshal${prop.schema.elementType.language.go!.discriminatorInterface}Array(*v)\n`;
+    } else if (prop.schema.language.go!.internalTimeType) {
+      unmarshaller += `\t\t\t\tvar aux ${prop.schema.language.go!.internalTimeType}\n`;
+      unmarshaller += '\t\t\t\terr = json.Unmarshal(*v, &aux)\n';
+      unmarshaller += `\t\t\t\t${receiver}.${prop.language.go!.name} = (*time.Time)(&aux)\n`;
+    } else {
+      unmarshaller += `\t\t\t\terr = json.Unmarshal(*v, &${receiver}.${prop.language.go!.name})\n`;
+    }
+    unmarshaller += '\t\t\t}\n';
+  }
+  unmarshaller += '\t\t}\n';
+  unmarshaller += '\t\tif err != nil {\n';
+  unmarshaller += '\t\t\treturn err\n';
+  unmarshaller += '\t\t}\n';
+  unmarshaller += '\t}\n';
+  if (!obj.language.go!.rootDiscriminator && parentType) {
+    unmarshaller += `\treturn json.Unmarshal(data, &${receiver}.${parentType.language.go!.name})\n`;
+  } else {
+    unmarshaller += '\treturn nil\n';
+  }
+  unmarshaller += '}\n\n';
+  // there's a corner-case where a derived type might not add any new fields (Cookiecuttershark).
+  // in this case skip adding the unmarshaller as it's not necessary and doesn't compile.
+  if (structDef.Properties && structDef.Properties.length > 0) {
```

Comment:
Wouldnt this be better at the beginning of the function? To avoid processing all of that work for this case where the unmarshaller isnt needed?

Reply:
Fixed